File: testdata/precision-suite-java/tp_pathtraversal.java
// Path traversal: a request parameter is used to build a filesystem path that
// is then read, with no canonicalization or allow-base check — CWE-22. A
// ../../etc/passwd value escapes the intended directory. A correct scanner
// fires TAINT-004.
package com.example.files;

import java.io.File;
import javax.servlet.http.HttpServletRequest;

public class FileServlet {

    // serve opens a file named directly by untrusted input — the vulnerability.
    public File serve(HttpServletRequest request) {
        String path = request.getParameter("file");
        return new File("/var/www/uploads/" + path); // nox-expect: TAINT-004
    }
}

File: testdata/precision-suite-kotlin/tp_pathtraversal.kt
// Path traversal: a user-controlled request parameter is used directly as a file
// path with no canonicalization or allow-base check, so `../../etc/passwd` reads
// outside the intended directory. This is CWE-22. A correct scanner fires
// TAINT-004. nox's Kotlin taint model sees `path` tainted by request.getParameter
// reach the FileInputStream(path) constructor sink with no File(...).name /
// normalize sanitizer on the path.
package com.example

import java.io.FileInputStream
import javax.servlet.http.HttpServletRequest

class Downloader {
    fun read(request: HttpServletRequest) {
        val path = request.getParameter("file")
        val stream = FileInputStream(path) // nox-expect: TAINT-004
        stream.close()
    }
}
